File: src/mrcv-objcourse.cpp
#include <mrcv/mrcv.h>
#include <mrcv/mrcv-common.h>

namespace mrcv
{
#ifdef _WIN32
    errno_t ObjCourse::initNN(const std::string pathToModel, const std::string pathToClasses)
#else
    error_t ObjCourse::initNN(const std::string pathToModel, const std::string pathToClasses)
#endif
	{
		
#ifdef _WIN32
        errno_t err = readClasses(pathToClasses);
#else
        error_t err = readClasses(pathToClasses);
#endif
		if (err == 0)
		{
			_network = cv::dnn::readNetFromONNX(pathToModel);
			if (_network.empty())
			{
				return ENETDOWN;
			}
			else
			{
				_network.setPreferableBackend(cv::dnn::DNN_BACKEND_DEFAULT);
				_network.setPreferableTarget(cv::dnn::DNN_TARGET_CPU);
			}
		}

		return err;
	}
    ObjCourse::ObjCourse(const std::string pathToModel, const std::string pathToClasses)
	{
		if (!initNN(pathToModel, pathToClasses))
		{
            writeLog("The neural network has been initiated successfully!");
            writeLog("Input width: " + std::to_string(_inputWidth));
            writeLog("Input height: " + std::to_string(_inputHeight));
		}
		else
		{
            writeLog("The neural network initialization ERROR!");
		}
	}
    ObjCourse::ObjCourse(const std::string pathToModel, const std::string pathToClasses, int width, int height)
	{
		_inputWidth = width;
		_inputHeight = height;

		if (!initNN(pathToModel, pathToClasses))
		{
			writeLog("The neural network has been initiated successfully!");
			writeLog("Input width: " + std::to_string(_inputWidth));
			writeLog("Input height: " + std::to_string(_inputHeight));
		}
		else
		{
			writeLog("The neural network initialization ERROR!");
		}
	}
    ObjCourse::ObjCourse(const std::string pathToModel, const std::string pathToClasses, int width, int height, float scoreThreshold, float nmsThreshold, float confidenceThreshold, float cameraAngle)
    {
        _inputWidth = width;
        _inputHeight = height;

        _scoreThreshold = scoreThreshold;
        _nmsThreshold = nmsThreshold;
        _confidenceThreshold = confidenceThreshold;

        _cameraAngle = cameraAngle;

        if (!initNN(pathToModel, pathToClasses))
        {
            writeLog("The neural network has been initiated successfully!");
            writeLog("Input width: " + std::to_string(_inputWidth));
            writeLog("Input height: " + std::to_string(_inputHeight));
        }
        else
        {
            writeLog("The neural network initialization ERROR!");
        }
    }
#ifdef _WIN32
    errno_t ObjCourse::readClasses(const std::string pathToClasses)
#else
    error_t ObjCourse::readClasses(const std::string pathToClasses)
#endif
	{
		std::ifstream classesFile(pathToClasses);
		std::string line;

		if (!classesFile)
		{
			writeLog("Failed to open classes names!");
			return ENOENT;
		}
		while (std::getline(classesFile, line))
		{
			_classes.push_back(line);
		}
		classesFile.close();

		return 0;
	}
    void ObjCourse::drawLabel(cv::Mat& img, std::string label, int left, int top)
    {
        // Отображение метки в боундинг боксе
        int baseline;
        cv::Size labelSize = cv::getTextSize(label, cv::FONT_HERSHEY_SIMPLEX, OBJCOURSE_FONT_SCALE, OBJCOURSE_THICKNESS, &baseline);
        top = std::max(top, labelSize.height);
        // Левый верхний угол
        cv::Point topLeftCorner = cv::Point(left, top);
        // Правый нижний угол
        cv::Point bottomRightCorner = cv::Point(left + labelSize.width, top + labelSize.height + baseline);        
        // Отрисовка
        cv::rectangle(img, topLeftCorner, bottomRightCorner, OBJCOURSE_BLACK, cv::FILLED);        
        cv::putText(img, label, cv::Point(left, top + labelSize.height), cv::FONT_HERSHEY_SIMPLEX, OBJCOURSE_FONT_SCALE, OBJCOURSE_YELLOW, OBJCOURSE_THICKNESS);
    }
	std::vector<cv::Mat> ObjCourse::preProcess(cv::Mat& img, cv::dnn::Net& net)
	{
		cv::Mat blob;
		cv::dnn::blobFromImage(img, blob, 1.0 / 255.0, cv::Size(_inputWidth, _inputHeight), cv::Scalar(), true, false);
		net.setInput(blob);
		std::vector<cv::Mat> outputs;
		net.forward(outputs, net.getUnconnectedOutLayersNames());
		return outputs;
	}
	cv::Mat ObjCourse::postProcess(cv::Mat& img, std::vector<cv::Mat>& outputs, const std::vector<std::string>& classNames)
	{
        // Начальная инициализация
        cv::Mat processedImage = img.clone();
        _classesIdSet.clear();
        _confidencesSet.clear();
        _boxesSet.clear();
        _classesSet.clear();

        std::vector<int> classIDs;
        std::vector<float> confidences;
        std::vector<cv::Rect> boxes;

        // Расчет коэффицентов ресайзинга
        float xFactor = img.cols / (float)_inputWidth;
        float yFactor = img.rows / (float)_inputHeight;

        float* data = (float*)outputs[0].data;

        const size_t dimensions = classNames.size() + 5;
        const size_t rows = 25200;        
        for (size_t i = 0; i < rows; ++i)
        {
            float confidence = data[4];
            // Пропускаем маловероятные объекты
            if (confidence >= _confidenceThreshold)
            {
                float* classesScores = data + 5;                
                cv::Mat scores(1, (int)classNames.size(), CV_32FC1, classesScores);                
                cv::Point classID;
                double maxClassScore;
                cv::minMaxLoc(scores, 0, &maxClassScore, 0, &classID);
                // Продолжаем если пройден вероятностный порог
                if (maxClassScore > _scoreThreshold)
                {
                    // Запоминаем класс ID
                    confidences.push_back(confidence);
                    classIDs.push_back(classID.x);                    
                    float cx = data[0];
                    float cy = data[1];
                    // Размер бокса
                    float w = data[2];
                    float h = data[3];
                    // Координаты боундинг бокса
                    int left = int((cx - 0.5 * w) * xFactor);
                    int top = int((cy - 0.5 * h) * yFactor);
                    int width = int(w * xFactor);
                    int height = int(h * yFactor);
                    // Запоминаем информацию о боксе
                    boxes.push_back(cv::Rect(left, top, width, height));
                }
            }
            data += dimensions;
        }

        std::vector<int> indexes;
        cv::dnn::NMSBoxes(boxes, confidences, _scoreThreshold, _nmsThreshold, indexes);

        mrcv::writeLog("POST PROCESS ===>");
        mrcv::writeLog("boxes.size(): " + std::to_string(boxes.size()));
        mrcv::writeLog("confidences.size(): " + std::to_string(confidences.size()));

        int bigestArea = INT_MIN;
        int bigestIndex = -1;
        int boxIndex = -1;

        for (size_t i = 0; i < indexes.size(); i++)
        {
            int idx = indexes[i];
            cv::Rect box = boxes[idx];

            boxIndex++;
            if (box.area() > bigestArea)
            {
                bigestIndex = boxIndex;
                bigestArea = box.area();
            }
        }

        if (bigestIndex > -1)
        {
            int idx = indexes[bigestIndex];
            cv::Rect box = boxes[idx];

            _boxesSet.push_back(box);
            _confidencesSet.push_back(confidences[idx]);
            _classesIdSet.push_back(classIDs[idx]);
            _classesSet.push_back(classNames[classIDs[idx]]);

            int left = box.x;
            int top = box.y;
            int width = box.width;
            int height = box.height;
            
            // Отрисовка боундинг бокса
            cv::rectangle(processedImage, cv::Point(left, top), cv::Point(left + width, top + height), OBJCOURSE_GREEN, 3 * OBJCOURSE_THICKNESS);
            
            // Получаем метку класса
            std::string label = cv::format("%.2f", confidences[idx]);
            label = classNames[classIDs[idx]] + ": " + label;
            if (OBJCOURSE_DRAW_LABEL)
            {
                // Отрисовка метки класса
                drawLabel(processedImage, label, left, top);
            }
        }

        return processedImage;
	}		
	std::string ObjCourse::getInfo(void)
	{
        std::string tmpString = "";
        for (size_t i = 0; i < _classesIdSet.size(); i++)
        {
            tmpString += _classesSet[i];
            tmpString += ": ";
            tmpString += std::to_string(_confidencesSet[i]);
            tmpString += "\n";
        }
        return tmpString;
	}
	cv::Mat ObjCourse::mainProcess(cv::Mat& img)
	{
        std::vector<cv::Mat> detections;
        detections = preProcess(img, _network);
        cv::Mat processResult = postProcess(img, detections, ObjCourse::_classes);
        std::vector<double> layersTimes;
        double freq = cv::getTickFrequency();
        ObjCourse::_inferenceTime = _network.getPerfProfile(layersTimes) / (float)freq;
        return processResult;
	}    
    int ObjCourse::findAngle(double resolution, double cameraAngle, int cx)
    {
        _cameraAngle = cameraAngle;
        return (int)((cx * _cameraAngle / resolution) - _cameraAngle / 2);
    }
    std::string ObjCourse::getTimeStamp()
    {
        using namespace std::chrono;
        auto now = system_clock::now();
        auto ms = duration_cast<milliseconds>(now.time_since_epoch()) % 1000;
        auto timer = system_clock::to_time_t(now);
        std::tm bt = *std::localtime(&timer);
        std::ostringstream oss;
        oss << std::put_time(&bt, "%d-%m-%Y %H:%M:%S"); // DD-MM-YYYY HH:MM:SS
        oss << '.' << std::setfill('0') << std::setw(3) << ms.count();
        return oss.str();
    }
    int ObjCourse::getObjectCount(cv::Mat frame)
    {
        cv::Mat img = mainProcess(frame);

        // Сохранить результат обработки на диск для отладки
        if (IS_DEBUG_LOG_ENABLED)
        {
            std::filesystem::path outputFile("files\\output.bmp");
            auto currentPath = std::filesystem::current_path();
            auto outputPath = currentPath / outputFile;
            cv::imwrite(outputPath.u8string(), img);
        }

        // Результаты работы детектора       
        std::vector<int> ids = getClassIDs();
        std::vector<float> confidences = getConfidences();
        std::vector<cv::Rect> boxes = getBoxes();
        std::vector<std::string> classes = getClasses();

        std::stringstream strIDs;
        for each (auto id in ids)
        {
            strIDs << std::to_string(id) << ";";
        }
        std::stringstream strConfs;
        for each (auto conf in confidences)
        {
            strConfs << std::to_string(conf) << ";";
        }
        writeLog("IDs: " + strIDs.str());
        writeLog("Confidence: " + strConfs.str());
        writeLog("Boxes: " + std::to_string(boxes.size()));
        
        return (int)boxes.size();
    }
    float ObjCourse::getObjectCourse(cv::Mat frame, double frameWidth, double cameraAngle)
    {
        cv::Mat img = mainProcess(frame);
        std::string timestamp = getTimeStamp();

        // Результаты работы детектора       
        std::vector<int> ids = getClassIDs();
        std::vector<float> confidences = getConfidences();
        std::vector<cv::Rect> boxes = getBoxes();
        std::vector<std::string> classes = getClasses();
        
        ///////////////////////////////////////////////////////////////////////
        // Поиск бокса цели с максимальной площадью
        ///////////////////////////////////////////////////////////////////////
        int bigestArea = INT_MIN;
        int bigestIndex = -1;
        int boxIndex = -1;

        for (auto b : boxes)
        {
            boxIndex++;
            if (b.area() > bigestArea)
            {
                bigestIndex = boxIndex;
                bigestArea = b.area();
            }
        }

        ///////////////////////////////////////////////////////////////////////
        // Расчет управления
        ///////////////////////////////////////////////////////////////////////        
        
        cv::Point center;
        cv::Point centerN;
        cv::Point centerP;
        cv::Point centerZ;

        std::string direction;
        int angle;
        
        if (boxes.size() > 0)
        {
            // Расчет центра бокса с обнаруженной целью
            center = (boxes[bigestIndex].br() + boxes[bigestIndex].tl()) * 0.5;

            // Угол между прицелом и целью
            angle = findAngle(frameWidth, cameraAngle, center.x);

            // Команда управления лево / право
            direction = center.x > frameWidth / 2 ? "RIGHT" : "LEFT";

            cv::Point boardBoxPt1;
            cv::Point boardBoxPt2;

            float SIGHT_WIDTH = 50; // Размер прицела
            
            // Координаты бокса прицела
            boardBoxPt1.x = (int)(img.cols / 2) - (int)SIGHT_WIDTH;
            boardBoxPt1.y = (int)(img.rows / 2) - (int)SIGHT_WIDTH;
            boardBoxPt2.x = (int)(img.cols / 2) + (int)SIGHT_WIDTH;
            boardBoxPt2.y = (int)(img.rows / 2) + (int)SIGHT_WIDTH;

            // Если цель находится в границах прицела - удерживаем курс
            //if ((boardBoxPt1.x <= center.x) && (center.x <= boardBoxPt2.x) &&
            //    (boardBoxPt1.y <= center.y) && (center.y <= boardBoxPt2.y))
            //{
            //    direction = "HOLD";
            //}

            // Алгоритм удержания цели только по оси абцисс
            if ((boardBoxPt1.x <= center.x) && (center.x <= boardBoxPt2.x))
                direction = "HOLD";

            // Время работы детектора
            std::stringstream ssTime;
            std::string inference;            

            // Строка инфорации
            std::string diagnosticInfo;

            // Время работы детектора
            ssTime.str(std::string()); // Очистка строкового стримера
            ssTime << std::fixed << std::setprecision(2) << getInference();
            inference = ssTime.str();
            
            // Отладочная информация
            diagnosticInfo = "CMD:\t(" + direction + ":" + std::to_string(angle) + ")" + "\tTIME: " + inference + "\t" + timestamp;
            writeLog(diagnosticInfo);

            return angle;
            
        }
        else
        {
            std::string diagnosticInfo = "TIME: " + timestamp + " NOT FOUND...";
            writeLog(diagnosticInfo);

            return 0.0;
        }

        return 0.0;
    }
}

File: python/mrcv/mrcv_objcourse.py
import cv2
import numpy as np
import torch
import onnxruntime as ort
import logging
import os
from datetime import datetime
import time
from typing import List, Tuple
from pathlib import Path

# Настройка логирования
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(message)s')
logger = logging.getLogger(__name__)

# Константы
OBJCOURSE_FONT_SCALE = 0.5
OBJCOURSE_THICKNESS = 2
OBJCOURSE_BLACK = (0, 0, 0)
OBJCOURSE_YELLOW = (0, 255, 255)
OBJCOURSE_GREEN = (0, 255, 0)
OBJCOURSE_DRAW_LABEL = True
IS_DEBUG_LOG_ENABLED = True

class ObjCourse:
    def __init__(self, path_to_model: str, path_to_classes: str, width: int = None, height: int = None,
                 score_threshold: float = None, nms_threshold: float = None,
                 confidence_threshold: float = None, camera_angle: float = None):
        self._input_width = width if width is not None else 640
        self._input_height = height if height is not None else 640
        self._score_threshold = score_threshold if score_threshold is not None else 0.5
        self._nms_threshold = nms_threshold if nms_threshold is not None else 0.45
        self._confidence_threshold = confidence_threshold if confidence_threshold is not None else 0.25
        self._camera_angle = camera_angle if camera_angle is not None else 60.0
        self._classes = []
        self._classes_id_set = []
        self._confidences_set = []
        self._boxes_set = []
        self._classes_set = []
        self._inference_time = 0.0

        # Инициализация нейронной сети
        if not self.init_nn(path_to_model, path_to_classes):
            logger.info("The neural network has been initiated successfully!")
            logger.info(f"Input width: {self._input_width}")
            logger.info(f"Input height: {self._input_height}")
        else:
            logger.error("The neural network initialization ERROR!")

    def init_nn(self, path_to_model: str, path_to_classes: str) -> int:
        # Чтение классов
        err = self.read_classes(path_to_classes)
        if err == 0:
            try:
                # Загрузка ONNX модели через onnxruntime
                self._session = ort.InferenceSession(path_to_model)
                self._input_name = self._session.get_inputs()[0].name
            except Exception as e:
                logger.error(f"Failed to load ONNX model: {e}")
                return 101  # Аналог ENETDOWN
        return err

    def read_classes(self, path_to_classes: str) -> int:
        try:
            with open(path_to_classes, 'r') as f:
                self._classes = [line.strip() for line in f if line.strip()]
            return 0
        except FileNotFoundError:
            logger.error("Failed to open classes file!")
            return 2  # Аналог ENOENT

    def draw_label(self, img: np.ndarray, label: str, left: int, top: int) -> None:
        font = cv2.FONT_HERSHEY_SIMPLEX
        label_size, baseline = cv2.getTextSize(label, font, OBJCOURSE_FONT_SCALE, OBJCOURSE_THICKNESS)
        top = max(top, label_size[1])
        top_left = (left, top)
        bottom_right = (left + label_size[0], top + label_size[1] + baseline)
        cv2.rectangle(img, top_left, bottom_right, OBJCOURSE_BLACK, cv2.FILLED)
        cv2.putText(img, label, (left, top + label_size[1]), font, OBJCOURSE_FONT_SCALE,
                    OBJCOURSE_YELLOW, OBJCOURSE_THICKNESS)

    def pre_process(self, img: np.ndarray) -> np.ndarray:
        # Преобразование изображения в формат, подходящий для модели
        img_resized = cv2.resize(img, (self._input_width, self._input_height))
        img_normalized = img_resized.astype(np.float32) / 255.0
        # Перевод в формат [N, C, H, W] для PyTorch/ONNX
        img_transposed = img_normalized.transpose(2, 0, 1)
        img_batch = np.expand_dims(img_transposed, axis=0)
        return img_batch

    def post_process(self, img: np.ndarray, outputs: np.ndarray, class_names: List[str]) -> np.ndarray:
        processed_image = img.copy()
        self._classes_id_set.clear()
        self._confidences_set.clear()
        self._boxes_set.clear()
        self._classes_set.clear()

        class_ids = []
        confidences = []
        boxes = []

        x_factor = img.shape[1] / self._input_width
        y_factor = img.shape[0] / self._input_height

        # Предполагается, что модель возвращает [batch, rows, dimensions]
        data = outputs[0]  # Первый выход модели
        rows = data.shape[0]
        dimensions = len(class_names) + 5

        for i in range(rows):
            confidence = data[i, 4]
            if confidence >= self._confidence_threshold:
                classes_scores = data[i, 5:]
                class_id = np.argmax(classes_scores)
                max_class_score = classes_scores[class_id]
                if max_class_score > self._score_threshold:
                    confidences.append(confidence)
                    class_ids.append(class_id)
                    cx, cy, w, h = data[i, :4]
                    left = int((cx - 0.5 * w) * x_factor)
                    top = int((cy - 0.5 * h) * y_factor)
                    width = int(w * x_factor)
                    height = int(h * y_factor)
                    boxes.append((left, top, width, height))

        # Применение NMS
        indices = cv2.dnn.NMSBoxes(boxes, confidences, self._score_threshold, self._nms_threshold)

        logger.info("POST PROCESS ==>")
        logger.info(f"boxes.size(): {len(boxes)}")
        logger.info(f"confidences.size(): {len(confidences)}")

        biggest_area = -float('inf')
        biggest_index = -1
        box_index = -1

        for i in indices.flatten():
            box_index += 1
            box = boxes[i]
            area = box[2] * box[3]
            if area > biggest_area:
                biggest_index = box_index
                biggest_area = area

        if biggest_index > -1:
            idx = indices[biggest_index]
            box = boxes[idx]
            self._boxes_set.append(box)
            self._confidences_set.append(confidences[idx])
            self._classes_id_set.append(class_ids[idx])
            self._classes_set.append(class_names[class_ids[idx]])

            left, top, width, height = box
            cv2.rectangle(processed_image, (left, top), (left + width, top + height),
                         OBJCOURSE_GREEN, 3 * OBJCOURSE_THICKNESS)

            if OBJCOURSE_DRAW_LABEL:
                label = f"{class_names[class_ids[idx]]}: {confidences[idx]:.2f}"
                self.draw_label(processed_image, label, left, top)

        return processed_image

    def get_info(self) -> str:
        return "\n".join(f"{cls}: {conf:.2f}" for cls, conf in zip(self._classes_set, self._confidences_set))

    def main_process(self, img: np.ndarray) -> np.ndarray:
        start_time = time.time()
        inputs = self.pre_process(img)
        outputs = self._session.run(None, {self._input_name: inputs})[0]
        processed_img = self.post_process(img, outputs, self._classes)
        self._inference_time = time.time() - start_time
        return processed_img

    def find_angle(self, resolution: float, camera_angle: float, cx: float) -> int:
        self._camera_angle = camera_angle
        return int((cx * camera_angle / resolution) - camera_angle / 2)

    def get_timestamp(self) -> str:
        now = datetime.now()
        return now.strftime("%d-%m-%Y %H:%M:%S.%f")[:-3]

    def get_object_count(self, frame: np.ndarray) -> int:
        img = self.main_process(frame)

        if IS_DEBUG_LOG_ENABLED:
            output_path = Path("files") / "output.bmp"
            output_path.parent.mkdir(exist_ok=True)
            cv2.imwrite(str(output_path), img)

        ids = self.get_class_ids()
        confidences = self.get_confidences()
        boxes = self.get_boxes()
        classes = self.get_classes()

        logger.info(f"IDs: {';'.join(map(str, ids))}")
        logger.info(f"Confidence: {';'.join(map(str, confidences))}")
        logger.info(f"Boxes: {len(boxes)}")

        return len(boxes)

    def get_object_course(self, frame: np.ndarray, frame_width: float, camera_angle: float) -> float:
        img = self.main_process(frame)
        timestamp = self.get_timestamp()

        ids = self.get_class_ids()
        confidences = self.get_confidences()
        boxes = self.get_boxes()
        classes = self.get_classes()

        biggest_area = -float('inf')
        biggest_index = -1
        box_index = -1

        for box in boxes:
            box_index += 1
            area = box[2] * box[3]
            if area > biggest_area:
                biggest_index = box_index
                biggest_area = area

        if boxes:
            box = boxes[biggest_index]
            left, top, width, height = box
            center = (left + width / 2, top + height / 2)
            angle = self.find_angle(frame_width, camera_angle, center[0])
            direction = "RIGHT" if center[0] > frame_width / 2 else "LEFT"

            sight_width = 50
            board_box_pt1 = (int(frame_width / 2 - sight_width), int(frame.shape[0] / 2 - sight_width))
            board_box_pt2 = (int(frame_width / 2 + sight_width), int(frame.shape[0] / 2 + sight_width))

            if board_box_pt1[0] <= center[0] <= board_box_pt2[0]:
                direction = "HOLD"

            diagnostic_info = f"CMD: ({direction}:{angle}) TIME: {self._inference_time:.2f} {timestamp}"
            logger.info(diagnostic_info)
            return angle

        logger.info(f"TIME: {timestamp} NOT FOUND...")
        return 0.0

    # Геттеры
    def get_class_ids(self) -> List[int]:
        return self._classes_id_set

    def get_confidences(self) -> List[float]:
        return self._confidences_set

    def get_boxes(self) -> List[Tuple[int, int, int, int]]:
        return self._boxes_set

    def get_classes(self) -> List[str]:
        return self._classes_set

    def get_inference(self) -> float:
        return self._inference_time